### PR TITLE
Beta geo expected probability no purchase

### PR DESCRIPTION
**Pull request by DylanZammit**
## Description
The linked issue below describes the motivation and derivation of the newly introduced method. Only 2 files have been updated:
- `pymc_marketing/clv/models/beta_geo.py`: Introduction of `BetaGeoModel.expected_probability_no_purchase`, which implements the derivation in aforementioned document.
- `docs/source/notebooks/bg_nbd.ipynb`: Includes an example of how to use it with some example customers.

I might require guidance on how to implement unit tests for this new method. Other unit tests for the BetaGeo model seem to compare the output results of similar methods with the ones given by the lifetimes library. However, there is no corresponding functionality for the new proposed method.

## Related Issue

- [ ] Closes #
- [x] Related to #1093 

## Checklist

- [x] Checked that [the pre-commit linting/style checks pass](https://docs.pymc.io/en/latest/contributing/python_style.html)
- [ ] Included tests that prove the fix is effective or that the new feature works 
- [x] Added necessary documentation (docstrings and/or example notebooks)
- [x] If you are a pro: each commit corresponds to a [relevant logical change](https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes)

## Modules affected

- [ ] MMM
- [x] CLV

## Type of change

- [x] New feature / enhancement
- [ ] Bug fix
- [ ] Documentation
- [ ] Maintenance
- [ ] Other (please specify):

----
📚 Documentation preview 📚: https://pymc-marketing--1094.org.readthedocs.build/en/1094/

### CODE REVIEW

File: pymc_marketing/clv/models/beta_geo.py [hunk 1]
```diff
@@ -437,6 +437,71 @@ def expected_probability_alive(
...
+        self,
+        t: int,
+        data: pd.DataFrame | None = None,
+    ) -> xarray.DataArray:
+        r"""Compute the probability a customer with history frequency, recency, and T
+        will have 0 purchases in the period (T, T+t].
+
+        The data parameter is only required for out-of-sample customers.
+
+        Adapted from Section 5.3 in Bruce Hardie's notes [1]_ for the special case when y = 0
```
Comment (juanitorduz) [inline pymc_marketing/clv/models/beta_geo.py:450]: Can you please link the number of the equation in the reference please?

**Comment by juanitorduz** [comment]
> I might require guidance on how to implement unit tests for this new method. 

Absolutely! Some ideas for this new method are:

- Test that the method does not fail by simply calling it on a fitted model as in https://github.com/pymc-labs/pymc-marketing/blob/e46f69016b5ec129e30c8a873ab8e0caf2694a70/tests/clv/models/test_beta_geo.py#L260

- Predict this on a user (or set of users) where this probability should be very high (close to one) and very low (close to zero)

**Comment by DylanZammit** [comment]
<p>Removed</p>

---

 <a href='https://app.reviewnb.com/pymc-labs/pymc-marketing/pull/1094/discussion/#comment-2423714654'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by DylanZammit** [comment]
> > I might require guidance on how to implement unit tests for this new method.
> 
> Absolutely! Some ideas for this new method are:
> 
> * Test that the method does not fail by simply calling it on a fitted model as in https://github.com/pymc-labs/pymc-marketing/blob/e46f69016b5ec129e30c8a873ab8e0caf2694a70/tests/clv/models/test_beta_geo.py#L260
> * Predict this on a user (or set of users) where this probability should be very high (close to one) and very low (close to zero)

Thank you! I will try to work on this between today and tomorrow. 

I have removed the redundant cell-block in the notebook and made reference to Equation 34 which I based the method off in the docstring of the new method. Appreciate the help and feedback, please let me know if there are any other suggested changes and I will review accordingly.

**Comment by DylanZammit** [comment]
@juanitorduz @ColtAllen 

Following unit tests added:
- assertion on expected output layout 
- frequent customers with large `t` should have probability of no purchases close to 0
- infrequent customers with small `t` should have probability of no purchases close to 1
- If `t` = 0, then the probability of no purchases should be 1

All tests passed.

File: pymc_marketing/clv/models/beta_geo.py (hunk 1 above)
Comment (ColtAllen) [reply pymc_marketing/clv/models/beta_geo.py]: Can we also add this reference to the [model class documentation](https://github.com/pymc-labs/pymc-marketing/blob/main/pymc_marketing/clv/models/beta_geo.py#L138)? Don't forget the numerical citation in the "First introduced by Fader. . . " paragraph.

File: pymc_marketing/clv/models/beta_geo.py [hunk 2]
```diff
@@ -437,6 +437,71 @@ def expected_probability_alive(
             "chain", "draw", "customer_id", missing_dims="ignore"
         )
 
+    def expected_probability_no_purchase(
```
Comment (ColtAllen) [inline pymc_marketing/clv/models/beta_geo.py:440]: Can we rename this to `expected_purchase_probability` for consistency with `ParetoNBDModel`? This method can be used for more than just the case of x=0.

**Review by ColtAllen — CHANGES_REQUESTED** [review]
Thanks for creating this PR @DylanZammit. I noticed the expression in the research note includes the model likelihood function, but it seems here you've simplified things and removed terms that would otherwise cancel out. If results are valid this will be a more performant solution, well done.

Aside from renaming the function and a few minor details in the notebook and docstrings, I don't have any major remarks.

**Comment by juanitorduz** [comment]
@DylanZammit I think this one is very close to the finish line. I think there are a couple of open comments from Colt, and then we are good to go. Please let us know if you need support :)

File: pymc_marketing/clv/models/beta_geo.py [hunk 3]
```diff
@@ -437,6 +439,71 @@ def expected_probability_alive(
...
+        t: int,
+        data: pd.DataFrame | None = None,
+    ) -> xarray.DataArray:
+        r"""Compute the probability a customer with history frequency, recency, and T
+        will have 0 purchases in the period (T, T+t].
+
+        The data parameter is only required for out-of-sample customers.
+
+        Adapted from Section 5.3, Equation 34 in Bruce Hardie's notes [1]_ for the special case when y = 0
+
```
Comment (ColtAllen) [inline pymc_marketing/clv/models/beta_geo.py:453]: Can we change _y_ to _frequency_ here for clarify?

**Comment by ColtAllen** [comment]
<p>Re-running the cell should clear it.</p>

---

 <a href='https://app.reviewnb.com/pymc-labs/pymc-marketing/pull/1094/discussion/#comment-2426594355'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by ColtAllen** [comment]
> @DylanZammit I think this one is very close to the finish line. I think there are a couple of open comments from Colt, and then we are good to go. Please let us know if you need support :)

I saw two typos in the notebook ("ot keep" and "deposits" instead of "spending"), but ReviewNB isn't allowing me to post any more comments for some reason. I'm also curious about that DIV/0 warning with the MAP fit, and noticed  an older version of `pytensor` in the watermark.

File: pymc_marketing/clv/models/beta_geo.py (hunk 3 above)
Comment (DylanZammit) [reply pymc_marketing/clv/models/beta_geo.py:453]: I have removed that part as suggested. This special case is indeed not mentioned in the research notes of Hardie. It is simply a special case I had derived. The derivation of this is linked in the PDF attached in the [original issue](https://github.com/pymc-labs/pymc-marketing/issues/1093) of this PR.

**Comment by DylanZammit** [comment]
Made all suggested changes:
* typos
* pytensor version (reinstalled the conda environment using the same command `conda env create -f environment.yml`)
* remove docstring reference of special case when y=0
* re-ran the cell that gave a warning (which removes the warning).

I have still not investigated the DIV/0 error, however, I suspect it is due to numerical issues. Please let me know if you have any further suggestions.

**Comment by ColtAllen** [comment]
> I have still not investigated the DIV/0 error, however, I suspect it is due to numerical issues. Please let me know if you have any further suggestions.

I'm not able to re-create the error on my 2023 M2 Macbook with identical `pymc` and `pytensor` versions; maybe it's a C++ compiler difference between your OS and mine. Anywho, if you add this cell to the top of the notebook and run it, then delete it before pushing, it'll hide all the warnings:

```
import warnings 
  
# Set warnings to be ignored 
warnings.filterwarnings('ignore')
```

**Review by ColtAllen — APPROVED** [review]
Good work on the notebook and augmenting existing research with a more performant function derivation! If you ever want to link your PDF to the docstring references feel free to open another PR.

**Pull request merged**